File: instrument_handle_barriers.pl
#!/usr/bin/env perl
use strict;
use warnings;

my $file = "cyberbrowser/third_party/quickjs/quickjs_handle_classes.h";
open my $fh, '<', $file or die $!;
my @lines = <$fh>;
close $fh;

my @out;
my $state = 0; # 0=normal, 1=collecting signature, 2=in body
my $sig_buf = "";
my $setter_type = "";
my $setter_param = "";
my $brace_depth = 0;

sub parse_sig {
    my ($s) = @_;
    if ($s =~ /void\s+set_[a-zA-Z0-9_]+\s*\(\s*(const\s+)?(GCValue|GCHandle)(?:\s*&\s*|\s+)([a-zA-Z0-9_]+)/s) {
        return ($2, $3);
    }
    return ("", "");
}

for my $line (@lines) {
    my $out_line = $line;

    if ($state == 0) {
        if ($line =~ /^\s*void\s+set_[a-zA-Z0-9_]+\s*\(/) {
            $sig_buf = $line;
            my $open = ($line =~ tr/(/(/);
            my $close = ($line =~ tr/)/)/);
            if ($open == $close) {
                ($setter_type, $setter_param) = parse_sig($sig_buf);
                if ($setter_type) {
                    print STDERR "ACTIVATE setter_type=$setter_type param=$setter_param\n";
                    $state = 2;
                    $brace_depth = ($sig_buf =~ tr/{/{/) - ($sig_buf =~ tr/}/}/);
                }
                $sig_buf = "";
            } else {
                $state = 1;
            }
        }
    } elsif ($state == 1) {
        $sig_buf .= $line;
        my $open = ($sig_buf =~ tr/(/(/);
        my $close = ($sig_buf =~ tr/)/)/);
        if ($open == $close) {
            ($setter_type, $setter_param) = parse_sig($sig_buf);
            if ($setter_type) {
                print STDERR "ACTIVATE setter_type=$setter_type param=$setter_param\n";
                $state = 2;
                $brace_depth = ($sig_buf =~ tr/{/{/) - ($sig_buf =~ tr/}/}/);
            } else {
                $state = 0;
            }
            $sig_buf = "";
        }
    } else {
        # in setter body
        my $open = ($line =~ tr/{/{/);
        my $close = ($line =~ tr/}/}/);
        $brace_depth += $open - $close;
        if ($line =~ /if \(p\)/) {
            print STDERR "IFP line='$line' depth=$brace_depth param='$setter_param'\n";
        }

        if ($line =~ /^\s*if \(p\) p->([a-zA-Z0-9_.\[\]]+)\s*=\s*\Q$setter_param\E\s*;\s*$/) {
            print STDERR "MATCH $setter_type $setter_param field=$1 depth=$brace_depth open=$open close=$close\n";
        }
        if ($line =~ /^\s*if \(p\) p->([a-zA-Z0-9_.\[\]]+)\s*=\s*\Q$setter_param\E\s*;\s*$/ && $open == 0 && $close == 0 && $brace_depth == 1) {
            print STDERR "EXPAND $setter_type $setter_param field=$1\n";
            my $field = $1;
            my $indent = $line =~ /^((?:    )*)/ ? $1 : "";
            my $barrier;
            if ($setter_type eq "GCHandle") {
                $barrier = "gc_write_barrier_for_heap_slot(&p->$field, $setter_param);";
            } else {
                $barrier = "gc_write_barrier_for_heap_slot(&p->$field, GC_VALUE_GET_HANDLE($setter_param));";
            }
            $out_line = "${indent}if (p) {\n${indent}    p->$field = $setter_param;\n${indent}    $barrier\n${indent}}\n";
            print STDERR "NEWLINE $out_line";
        }

        if ($brace_depth <= 0) {
            print STDERR "CLOSE setter depth=$brace_depth line='$line' open=$open close=$close";
            $state = 0;
            $setter_type = "";
            $setter_param = "";
        }
    }

    push @out, $out_line;
}

open my $out, '>', $file or die $!;
print $out @out;
close $out;
print "Done\n";

File: instrument_handle_barriers.py
#!/usr/bin/env python3
"""Instrument GC-reference-writing setters in quickjs_handle_classes.h
   with write barriers using this->handle_ as the source."""
import re
from pathlib import Path

FILE = Path("cyberbrowser/third_party/quickjs/quickjs_handle_classes.h")
text = FILE.read_text(encoding="utf-8")

# We transform setters of the form:
#   void set_xxx(const GCValue& val) {
#       SomeType* p = get_ptr();
#       if (p) p->field = val;
#   }
# or with GCValue v, or GCHandle h/handle/val.
#
# We only touch bodies that contain a single assignment guarded by if (p).
# More complex methods are left for manual review.

def repl_gvalue(m):
    body = m.group(0)
    # Find the parameter name: const GCValue& NAME  or  GCValue NAME
    sig = m.group(1)
    # The full match includes the braces; we need the assignment line.
    # Extract the field expression after "p->" and before " = <param>;"
    ma = re.search(r'if \(p\) p->([a-zA-Z0-9_.\[\]]+)\s*=\s*([a-zA-Z0-9_]+)\s*;', body)
    if not ma:
        return body
    field = ma.group(1)
    param = ma.group(2)
    # Validate parameter name appears in signature
    if param not in sig:
        return body
    new_body = body[:ma.start()] + \
        f"if (p) {{\n            p->{field} = {param};\n            gc_write_barrier(handle_, GC_VALUE_GET_HANDLE({param}));\n        }}" + \
        body[ma.end():]
    return new_body

def repl_ghandle(m):
    body = m.group(0)
    sig = m.group(1)
    ma = re.search(r'if \(p\) p->([a-zA-Z0-9_.\[\]]+)\s*=\s*([a-zA-Z0-9_]+)\s*;', body)
    if not ma:
        return body
    field = ma.group(1)
    param = ma.group(2)
    if param not in sig:
        return body
    new_body = body[:ma.start()] + \
        f"if (p) {{\n            p->{field} = {param};\n            gc_write_barrier(handle_, {param});\n        }}" + \
        body[ma.end():]
    return new_body

# Match methods with a single-line if assignment body.
# We use a regex that spans from "void set_...(" to the closing brace,
# but stops at first "}". This works for simple bodies only.
pat_gvalue = re.compile(
    r'void set_([a-zA-Z0-9_]+)\(((?:const\s+)?GCValue(?:\s*&\s*|\s+)([a-zA-Z0-9_]+))\)\s*\{[^}]*if \(p\) p->[a-zA-Z0-9_.\[\]]+\s*=\s*\3\s*;[^}]*\}',
    re.DOTALL
)
pat_ghandle = re.compile(
    r'void set_([a-zA-Z0-9_]+)\((GCHandle\s+([a-zA-Z0-9_]+))\)\s*\{[^}]*if \(p\) p->[a-zA-Z0-9_.\[\]]+\s*=\s*\3\s*;[^}]*\}',
    re.DOTALL
)

# Apply replacements until stable (avoid overlapping issues)
old = None
while old != text:
    old = text
    text = pat_gvalue.sub(repl_gvalue, text)
    text = pat_ghandle.sub(repl_ghandle, text)

FILE.write_text(text, encoding="utf-8")
print("Done")
